agent/internal/container/container.go

```diff
@@ -0,0 +1,428 @@
+}
+
+// Signal asynchronously delivers the signal. Delivery failures are surfaced in logs.
+func (c *Container) Signal(ctx context.Context, s syscall.Signal) {
+	select {
+	case c.signals <- s:
+	case <-ctx.Done():
+		c.log.Warnf("ignoring signal on container due to cancellation: %v", s)
+	default:
+		c.log.Warnf("ignoring signal on container with too many pending signals: %v", s)
+	}
+}
+
+// Wait until the container exits. Always returns a ContainerExit unless canceled by Detach.
+func (c *Container) Wait() *aproto.ContainerStateChanged {
+	<-c.done
+	return c.exit
+}
+
+// run the container. If the context is canceled, the container is detached (as is) and
+// the context error is returned. If the container is killed, the container is cleaned up if
+// necessary and a termination with a stack trace or kill result is returned. If the container is
+// started successful, both the termination and error returns will be nil.
+func (c *Container) run(parent context.Context) (err error) {
+	c.log.Trace("entering run")
+	ctx, cancel := context.WithCancel(parent) // run-scoped cancellable context.
+	defer cancel()
+
+	killed := atomic.NewBool(false)
+	defer func() {
+		if killed.Load() {
+			if err != nil {
+				c.log.WithError(err).Debug("exit of container killed before run")
+			}
+			c.log.Trace("converting error to container aborted, since we were killed")
+			err = ErrKilledBeforeRun
+		}
+	}()
+
+	// Until the container image is pulled, just catch kill signals, note it, and cancel run.
+	c.log.Trace("launching signal-to-context shimmer")
+	siggroup := waitgroupx.WithContext(ctx)
+	siggroup.Go(func(ctx context.Context) {
+		defer siggroup.Cancel()
+		for {
+			select {
+			case signal := <-c.signals:
+				switch signal {
+				case syscall.SIGKILL:
+					c.log.Tracef("signal %s, canceling run-scoped context", signal)
+					killed.Store(true)
+					cancel()
+					return
+				default:
+					c.log.Warnf("ignoring signal other than SIGKILL %s before running", signal)
+				}
+			case <-ctx.Done():
+				c.log.Trace("signal-to-context shimmer exited")
+				return
+			}
+		}
+	})
+
+	c.log.Trace("pulling image")
+	if err = c.transition(ctx, cproto.Pulling, nil, nil); err != nil {
+		return err
+	}
+	if err = c.docker.PullImage(ctx, docker.PullImage{
+		Name:      c.spec.RunSpec.ContainerConfig.Image,
+		Registry:  c.spec.PullSpec.Registry,
+		ForcePull: c.spec.PullSpec.ForcePull,
+	}, c.shimDockerEvents()); err != nil {
+		return err
+	}
+
+	c.log.Trace("creating container, copying files, initializing watches, etc")
+	if err = c.transition(ctx, cproto.Starting, nil, nil); err != nil {
+		return err
+	}
+	dockerID, err := c.docker.CreateContainer(ctx, c.spec.RunSpec, c.shimDockerEvents())
+	if err != nil {
+		return err
+	}
+	remove := c.spec.RunSpec.HostConfig.AutoRemove
+	c.spec = nil // Evict the spec from memory due to their potential memory consumption.
+
+	// Ensure we don't miss kill signals that received after RunContainer but before we stopped
```

Comment:
these signals will be handled by the loop in `.wait`, no? I am not sure where this "shimming to context cancellations" is...

Reply:
killed is set to false